feat: Added new date wrapper

petertonysmith94 opened the pull request:
Closes #1573 

### Changes made
- Added `DateTime` class, which enables conversion to and from timestamps based in the following formats **Tai64** and **Unix**.
- Added documentation around the conversion of common time formats.
- Removed the `tai64` package dependency.

### Known limitations

There were two implementations available with the codebase:

 - Package `tai64` ([npm](https://www.npmjs.com/package/tai64))
 - Custom implementation using `BigInt` and a constant "leap second" offset.
 
 This leap second offset will need to be [updated](https://data.iana.org/time-zones/data/leap-seconds.list), this is how the the `tai64` package handles this [implementation](https://github.com/hl2/tai64/blob/master/src/leapSeconds.ts).
 
 I have gone with the custom implementation as it's more lightweight, favouring consolidating to a single solution and
 using `BigInt` to follow the general goals of this community (#1542).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/FuelLabs/fuels-ts?pullRequest=1627) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/FuelLabs/fuels-ts?pullRequest=1627) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: danielbate<br/>:white_check_mark: Torres-ssf<br/>:white_check_mark: petertonysmith94<br/>:x: nedsalk<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/FuelLabs/fuels-ts?pullRequest=1627) it.</sub>

arboleya commented:
We're about to freeze `salamander` RC for final review and integration.

Could you please base this PR against `master`?

petertonysmith94 commented:
@FuelLabs/Frontend 

Was hoping for a review/sign off from the FE team perspective, for the following API. Does it meet you requirements?
---

The following instantiation methods are available:
```ts
const tai64: IDateTime = DateTime.fromTai64('4611686020108779312');
const unixSeconds: IDateTime = DateTime.fromUnixSeconds(1681391398);
const unixMilliseconds: IDateTime = DateTime.fromUnixMilliseconds(1681391398000);
```
Note: actively removed instantiation via the constructor (to avoid confusions between timestamps).


---

Provided are the following methods of conversion and fully date compatible:

```ts
const date: IDateTime = DateTime.fromTai64('4611686020108779312');
date.toTai64() // (string) '4611686020108779312'
date.toUnixMilliseconds() // (number) 1681391398000
date.toUnixSeconds() // (number) 1681391398
```

Dhaiwat10 commented:
@petertonysmith94 looks like this PR might need a rebase. 😅

arboleya commented:
> @petertonysmith94 looks like this PR might need a rebase. 😅

It's a minor conflict; a simple merge should suffice. 😌

nedsalk commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+# Date Time
```
```suggestion
# DateTime
```

nedsalk commented on packages/utils/src/utils/date-time.ts:
```diff
@@ -0,0 +1,212 @@
+/**
+ * Interface for the DateTime class.
+ */
+export interface IDateTime extends Date {
```
There is no need to create an interface if there's only one implementer of it (`DateTime`). Wherever we need just the type of the class, we can do `import type { DateTime } from '...'`, and if we need the actual class we do the standard `import { DateTime } from '...'`. I'm requesting changes for the deletion of the interface.

nedsalk replied:
Btw the rest seems good! :+1:

danielbate replied:
> There is no need to create an interface if there's only one implementer of it

I disagree with this as there may be times where something is designed with extensibility in mind, but right now there may only be one implementer. Giving less overhead to the extension.

However context is important, and I agree here the interface is unnecessary. We're in the realm of `AbstractAddress`.

petertonysmith94 replied:
[23778c70a32127e963c7dfb2e98749bb42901fdb](https://github.com/FuelLabs/fuels-ts/pull/1627/commits/23778c70a32127e963c7dfb2e98749bb42901fdb)

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+To allow for easier manipulation of date and time, the SDK exports the `DateTime` class, which is a wrapper around the `Date` class ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)). Below we will go over the methods of instantiation, utility functions, and time formats.
+
+Why is this useful?
+Internally the transactions and other time/date assets are encoded using the **TAI64** format. We return a `IDateTime` interface, to allow of easier conversion and formatting between the two formats.
```
Since here is the first mention of 'tai64' on this page, can you make the first occurrence of the term a link to the section where you have explained the format down below on the same page?

petertonysmith94 replied:
That seems wise!

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+To allow for easier manipulation of date and time, the SDK exports the `DateTime` class, which is a wrapper around the `Date` class ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)). Below we will go over the methods of instantiation, utility functions, and time formats.
+
+Why is this useful?
```
Maybe make this a subheading as well?

petertonysmith94 replied:
Will remove - redundant

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+## API
+
+We have a host of static method for **instantiation** of our `DateTime` class.
```
Nit sorry - typo 
```suggestion
We have a host of static methods for **instantiation** of our `DateTime` class.
```

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+`fromUnixMilliseconds` is a _static_ method, that allows the creation of `DateTime` class from a UNIX Milliseconds number.
+
+`toUnixMilliseconds` is a _instance_ method, that allows the conversion of a `DateTime` class to a `UNIX` number in milliseconds.
```
```suggestion
`toUnixMilliseconds` is an _instance_ method, that allows the conversion of a `DateTime` class to a `UNIX` number in milliseconds.
```

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+`fromUnixSeconds` is a _static_ method, that allows the creation of `DateTime` class from a UNIX Seconds number.
+
+`toUnixSeconds` is a _instance_ method, that allows the conversion of a `DateTime` class to a `UNIX` number in seconds.
```
```suggestion
`toUnixSeconds` is an _instance_ method, that allows the conversion of a `DateTime` class to a `UNIX` number in seconds.
```

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+### Date
+
+The `DateTime` class extends the functionality of the `Date` object (see here), so all method are available for your usages.
```
Did you mean to make 'see here' a link?

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+### TAI Format
+
+TAI stands for _Temps Atomique International_ and is the current international real-time standard [Source](https://cr.yp.to/libtai/tai64.html).
```
```suggestion
TAI stands for _Temps Atomique International_ and is the current international real-time standard. [Source](https://cr.yp.to/libtai/tai64.html).
```

petertonysmith94 replied:
There is a full-stop at the end of the sentence, will skip this one 😄

Dhaiwat10 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+TAI stands for _Temps Atomique International_ and is the current international real-time standard [Source](https://cr.yp.to/libtai/tai64.html).
+
+We use **TAI64** is a 64-bit integer representing the number of nanoseconds since the epoch.
```
```suggestion
`TAI64` is a 64-bit integer representing the number of nanoseconds since the epoch.
```

I would also recommend replacing all `**TAI64**` with `TAI64` (<code> markdown) instead

petertonysmith94 replied:
I concur, that is better 😍

Dhaiwat10 reviewed (changes requested):
Awesome job @petertonysmith94! Left some possible improvements for the docs and some nits here and there. Feel free to incorporate them as you deem necessary :)

danielbate commented on packages/account/src/providers/scalars.test.ts:
Really good to see the scalar introduction, just referencing #1572 to extend their use there.

danielbate commented on packages/account/src/providers/scalars.test.ts:
```diff
@@ -0,0 +1,217 @@
+import { Tai64Scalar, UnixMillisecondScalar } from './scalars';
+
+/**
+ * @group node
```
```suggestion
 * @group node
 * @group browser
```

danielbate commented on packages/account/src/providers/scalars.ts:
Might be nice to have a `scalars` directory, and export the lot form and `index` file. Then we can also split the test files out.

danielbate commented on packages/utils/src/utils/date-time.test.ts:
```diff
@@ -0,0 +1,130 @@
+};
+
+/**
+ * @group node
```
```suggestion
 * @group node
 * @group browser
```

danielbate commented on packages/utils/src/utils/date-time.ts:
Only concern here as I mentioned above is the mix of objects and utility functions.

petertonysmith94 replied:
I have unexported the utility functions, I believe they still hold value for readability reasons, so have kept them within the `date-time` file.

petertonysmith94 replied:
Refactored to using only using the `DateTime` within the codebase also.

danielbate commented on pnpm-lock.yaml:
Lock file has changed but I can't see any package changes? 🤔  Maybe just exclude this file.

danielbate reviewed (changes requested):
Great work @petertonysmith94 👍🏻 

My main thought here is I think we should choose between whether we are using the `DateTime` wrapper for state and conversions, or whether we are just going to use utility functions. I like the wrapper as we are extending `Date` regardless and supports the scalars. However in that case we should not export the utilities IMO. Thoughts?

petertonysmith94 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+<<< @/../../docs-snippets/src/guide/types/date-time.test.ts#date-object-methods{ts:line-numbers}
+
+## Formats
```
For me, that would just be rehashing what has been covered above, and has bloat potential, will leave unless theres a strong sentiment to this change 😄

danielbate replied:
Sorry I mean you should also get rid of what you say above. So rather than splitting up instantiating and formats, just have a subheading for the format and then how to instantiate it.

IMO that would be bless bloated than how it currently is. It would make sense to put everything UNIX related together and everything TAI related together.

danielbate replied:
```md
# DateTime

To allow for easier...

## Instantiating a `DateTime`

We have a host of...

### TAI64

TAI stands for...

Instantiating TAI64...

### UNIX

UNIX intro

Instantiating UNIX...

### Date

Date intro

Instantiating from date...
```

Formats as a sub para seems unnecessary and flows oddly.

petertonysmith94 replied:
I personally believe the serperation between the two, gives better visibility to the API, which is what you want surely?

I do agree that the flow isn't ideal, but went with an API first approach with documentation, which I think is more valuable for the user IMO

petertonysmith94 commented on apps/docs/src/guide/types/date-time.md:
```diff
@@ -0,0 +1,60 @@
+
+`fromUnixMilliseconds` is a _static_ method, that allows the creation of `DateTime` class from a UNIX Milliseconds number.
+
+`toUnixMilliseconds` is a _instance_ method, that allows the conversion of a `DateTime` class to a `UNIX` number in milliseconds.
```
Resolved [here](https://github.com/FuelLabs/fuels-ts/pull/1627/files/432fd8ad72ec60428b885d07604a2acb21b58653#r1498308627)

Torres-ssf commented on packages/account/src/providers/scalars.ts:
```diff
@@ -0,0 +1,50 @@
+  },
+});
+
+export const Tai64Scalar = new GraphQLScalarType<IDateTime, string>({
+  name: 'Tai64',
+  description: 'Tai64 custom scalar type',
+  serialize(value: unknown): string {
+    if (value instanceof DateTime) {
+      return value.toTai64();
+    }
```
Is this custom Scalar being used? How it communicates with `codegen` ?

I am asking this because after some investigation I didn't find a way to use custom scalars with `codegen` within a Node Graphql client application. 

Since we have an [issue](https://github.com/FuelLabs/fuels-ts/issues/1572) related to this, I am curious if this is possible or if we need to close this issue.

petertonysmith94 replied:
Hey @Torres-ssf,

It is not being used (assumed that would come later), I just wanted to show that the `DateTime` implemented, does meet the requirements, such that, it can be used as a GraphQL Scalar.

There been a lot of talk about custom scalars for client-side based applications, however, can't find much that isn't Apollo specific. Potentially we could draw inspiration from a package written for Apollo [apollo-client-transformers](https://github.com/cult-of-coders/apollo-client-transformers/tree/master).

References #1572

petertonysmith94 replied:
Decided to remove the scalars.
[6dd0716851556358d2dd0b6125642eec86d2c605](https://github.com/FuelLabs/fuels-ts/commit/6dd0716851556358d2dd0b6125642eec86d2c605)

Torres-ssf commented on apps/docs-snippets/src/guide/types/date-time.test.ts:
```diff
@@ -0,0 +1,98 @@
+describe(__filename, () => {
+  it('should be able to be created from multiple sources', () => {
+    // #region create-from-multiple-sources
+    // #context import type { IDateTime } from 'fuels';
+    // #context import { DateTime } from 'fuels';
```
Exactly @danielbate.

We should add imports to code snippets using the following approach:

```ts
// #import { IDateTime, DateTime };
```

petertonysmith94 commented on apps/docs-snippets/src/guide/testing/tweaking-the-blockchain.test.ts:
```diff
@@ -15,7 +15,7 @@ test('produceBlocks with custom timestamp docs snippet', async () => {
   const lastBlockNumber = latestBlock.height;
   // #region Provider-produceBlocks-custom-timestamp
-  const lastBlockTimestamp = fromTai64ToUnix(latestBlock.time);
+  const lastBlockTimestamp = DateTime.fromTai64(latestBlock.time).getTime();
```
You are semi-correct!
`getTime` === `toUnixMilliseconds`

I have changed this `toUnixMilliseconds`

petertonysmith94 commented on apps/docs-snippets/src/guide/types/date-time.test.ts:
```diff
@@ -0,0 +1,98 @@
+import { DateTime } from 'fuels';
+
+/**
+ * @group node
```
None of the other doc-snippets are running in the brower currently, so going to miss this one

danielbate reviewed (approved):
LGTM 👍🏻  Just one [comment](https://github.com/FuelLabs/fuels-ts/pull/1627#discussion_r1500797601) around docs but see what others think around the additional verbosity.

Torres-ssf commented on packages/account/codegen.json:
```diff
@@ -22,7 +22,8 @@
           "Salt": "string",
-          "Nonce": "string"
+          "Nonce": "string",
+          "Tai64Timestamp": "string"
```
@petertonysmith94 I believe we can also remove this since it has no use for the `graphql-codegen`

petertonysmith94 replied:
Hey @Torres-ssf I'd disagree with this one. Without the `Tai64Timestamp` we would get a type of `any` without our generated operations. We know the `Tai64Timestamp` is a `string`, so why not type hint it?

Torres-ssf replied:
@petertonysmith94 You're right. 

I did forget that `Tai64Timestamp` is already a Scalar defined on the GraphQL server. 

Thanks for clarifying that.

petertonysmith94 replied:
@Torres-ssf no worries at all, thank you for your review!

arboleya commented on packages/utils/src/utils/date-time.test.ts:
```diff
@@ -0,0 +1,95 @@
+import { DateTime } from './date-time';
+
+const { now, nowPlusOneSecond } = {
```
NIT obs — this is a funny one; why not just use plain declarations? 😄

You'd have saved one line plus a few chars, btw.

```ts
const now = {
  tai64: '4611686020108779339',
  unixSeconds: 1681391398,
  unixMilliseconds: 1681391398000,
  isoString: '2023-04-13T13:09:58.000Z',
}

const nowPlusOneSecond = {
  tai64: '4611686020108779340',
  unixSeconds: 1681391399,
  unixMilliseconds: 1681391399000,
  isoString: '2023-04-13T13:09:59.000Z',
}
```

arboleya replied:
You didn't need to, but thanks!

arboleya reviewed (approved):
Thank you for yet another contribution! 🙌

petertonysmith94 commented:
Hey @Dhaiwat10 and @nedsalk,

I've _hopefully_ addressed all the comments now, thank you for the eyes on! 👀  Please could you take another peek when you have a minute or two, I would be eternally grateful 🙏🏼

petertonysmith94 commented:
<img width="951" alt="Screenshot 2024-02-26 at 21 03 40" src="https://github.com/FuelLabs/fuels-ts/assets/16990131/28df1d72-7a2c-40d0-a975-8735d16e341c">

Anyone with write access that would/could - pretty please 🍒  merge this into `master` 🦖

arboleya commented:
@petertonysmith94 Thanks again for another PR! 1️⃣➕
